radix-engine/src/track/track.rs

```diff
@@ -114,22 +110,106 @@ impl TrackedSubstateKey {
             TrackedSubstateKey::New(substate)
             | TrackedSubstateKey::WriteOnly(Write::Update(substate))
             | TrackedSubstateKey::ReadOnly(ReadOnly::Existent(substate))
-            | TrackedSubstateKey::ReadAndWrite(_, Write::Update(substate)) => Some(substate),
+            | TrackedSubstateKey::ReadExistAndWrite(_, Write::Update(substate))
+            | TrackedSubstateKey::ReadNonExistAndWrite(substate) => Some(substate),
             TrackedSubstateKey::WriteOnly(Write::Delete)
-            | TrackedSubstateKey::ReadAndWrite(_, Write::Delete)
-            | TrackedSubstateKey::ReadOnly(ReadOnly::NonExistent) => None,
+            | TrackedSubstateKey::ReadExistAndWrite(_, Write::Delete)
+            | TrackedSubstateKey::ReadOnly(ReadOnly::NonExistent)
+            | TrackedSubstateKey::Garbage => None,
         }
     }
 
-    pub fn get_substate(&self) -> Option<&IndexedScryptoValue> {
+    pub fn get(&self) -> Option<&IndexedScryptoValue> {
         match self {
             TrackedSubstateKey::New(substate)
             | TrackedSubstateKey::WriteOnly(Write::Update(substate))
             | TrackedSubstateKey::ReadOnly(ReadOnly::Existent(substate))
-            | TrackedSubstateKey::ReadAndWrite(_, Write::Update(substate)) => Some(&substate.value),
+            | TrackedSubstateKey::ReadExistAndWrite(_, Write::Update(substate))
+            | TrackedSubstateKey::ReadNonExistAndWrite(substate) => Some(&substate.value),
             TrackedSubstateKey::WriteOnly(Write::Delete)
-            | TrackedSubstateKey::ReadAndWrite(_, Write::Delete)
-            | TrackedSubstateKey::ReadOnly(ReadOnly::NonExistent) => None,
+            | TrackedSubstateKey::ReadExistAndWrite(_, Write::Delete)
+            | TrackedSubstateKey::ReadOnly(ReadOnly::NonExistent)
+            | TrackedSubstateKey::Garbage => None,
+        }
+    }
+
+    pub fn set(&mut self, value: IndexedScryptoValue) {
+        match self {
+            TrackedSubstateKey::Garbage => {
+                *self = TrackedSubstateKey::WriteOnly(Write::Update(RuntimeSubstate::new(value)));
```

Comment:
Can it be `ReadAndWrite` here?

Reply:
Nope, this should never be `ReadAndWrite` as `Garbage` is only created from `WriteOnly`